File: server/schemas/gfa.py
from enum import IntEnum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel

segment_optional_fields = ["LN", "RC", "FC", "KC", "SH", "UR"]
link_optional_fields = ["MQ", "NM", "RC", "FC", "KC", "ID"]


class GFA_ELEMENT(IntEnum):
    SEGMENT = 0
    LINK = 1
    PATH = 2


class GfaElement(BaseModel):
    # TODO: change any to union of more specific types
    type: str
    name: str
    optionals: Optional[Dict[str, Any]] = None


class GfaPath(GfaElement):
    type = "path"
    segment_names: List[str]
    # TODO: overlaps


class GfaSegment(GfaElement):
    type = "segment"
    sequence: str
    paths: List[GfaPath]


class GfaLink(GfaElement):
    type = "link"
    from_segment: str
    from_orient: str
    to_segment: str
    to_orient: str
    paths: List[GfaPath]
    # TODO: overlap


class Gfa(BaseModel):
    segments: List[GfaSegment]
    links: List[GfaLink]
    paths: List[GfaPath]


class GfaInfo(BaseModel):
    # TODO:  add more
    n_segments: int
    n_links: int
    n_paths: int


class GfaHist(BaseModel):
    hist: List[float]
    bin_edges: List[float]

File: src/types/gfa.ts
// TODO: containment

interface Gfa {
  readonly header?: GfaHeader;
  readonly segments: GfaSegment[];
  readonly links: GfaLink[];
  readonly paths: GfaPath[];
}

interface GfaElement {
  type: 'segment' | 'link' | 'path';
  readonly name: string;
  // TODO: make sure that value string | number fits all scenarios
  readonly optionals?: Record<GfaOptional, any>;
}

interface GfaHeader {}

interface GfaSegment extends GfaElement {
  type: 'segment';
  readonly sequence: string;
  readonly paths: GfaPath[];
}

interface GfaLink extends GfaElement {
  type: 'link';
  readonly from_segment: string;
  readonly from_orient: string;
  readonly to_segment: string;
  readonly to_orient: string;
  readonly paths: GfaPath[];
  // TODO: overlap
}

interface GfaPath extends GfaElement {
  readonly type: 'path';
  readonly segment_names: string[];
  // TODO: overlaps
}

interface GfaInfo {
  readonly n_segments: number;
  readonly n_links: number;
  readonly n_paths: number;
}

interface GfaHist {
  readonly hist: number[];
  readonly bin_edges: number[];
}

type GfaOptional = 'VN' | 'LN' | 'RC' | 'FC' | 'KC' | 'SH' | 'UR' | 'MQ' | 'NM' | 'ID';
type GfaFeature = GfaSegment | GfaLink;

export const getSegmentLength = (segment: GfaSegment) => {
  if (segment.optionals && segment.optionals.LN) {
    return segment.optionals.LN;
  } else {
    return segment.sequence.length;
  }
};

export type {
  GfaElement,
  GfaHeader,
  GfaSegment,
  GfaLink,
  GfaOptional,
  GfaPath,
  GfaInfo,
  GfaHist,
  GfaFeature,
};
export default Gfa;
